File: js/packages/serializer/models/SerializerBuffer.js
import { SmartBuffer } from 'smart-buffer';

class SerializerBuffer {
    #buffer = new SmartBuffer()

    #bitOffset = -1
    #bitBuffer = 0
    #bitCount = 0

    get length() {
        return this.#buffer.length
    }

    #flushBits() {
        if (this.#bitCount === 0) return

        this.#buffer.writeUInt8(this.#bitBuffer, this.#bitOffset)
        this.#bitCount = 0
        this.#bitBuffer = 0
    }

    #writeLSBits(value, bitLength) {
        if (this.#bitCount === 0) {
            // Reserve space
            this.#bitOffset = this.#buffer.writeOffset
            this.#buffer.writeUInt8(0)
        }

        const mask = (1 << bitLength) - 1
        const masked = value & mask
        const shifted = masked << this.#bitCount

        this.#bitCount += bitLength
        this.#bitBuffer |= shifted
    }

    // Will truncate all significant bits above the bitLength
    writeTruncated(value, bitLength) {
        while (bitLength > 0) {
            const available = 8 - this.#bitCount
            if (bitLength < available) {
                this.#writeLSBits(value, bitLength)
                break;
            }

            this.#writeLSBits(value, available)
            this.#flushBits()

            value >>= available
            bitLength -= available
        }
    }

    writeSigned(value, bitLength) {
        // Bit shifts return 32-bit SIGNED integers so we cannot decode 32-bit unsigned values.
        // Technically we could allow 32-bit signed bit fields only, but that would be weird
        // Just use a regular Int/UInt
        if (bitLength > 31) throw new Error('Bit fields cannot be larger that 31 bits')

        // Here we could use bit shifts to calculate the range,
        // but I prefer to stay consistent with the unsigned implementation where we cannot
        const minValue = -Math.pow(2, bitLength - 1)
        const maxValue = Math.pow(2, bitLength - 1) - 1

        if (value < minValue || value > maxValue)
            throw new Error('Value out of range')

        this.writeTruncated(value, bitLength)
    }

    writeUnsigned(value, bitLength) {
        if (bitLength > 31) throw new Error('Bit fields cannot be larger that 31 bits')

        const maxValue = Math.pow(2, bitLength) - 1

        if (value < 0 || value > maxValue)
            throw new Error('Value out of range')

        this.writeTruncated(value, bitLength)
    }

    writeByte(byte) {
        this.#buffer.writeInt8(byte)
    }

    writeUByte(uByte) {
        this.#buffer.writeUInt8(uByte)
    }

    writeShort(short) {
        this.#buffer.writeInt16LE(short)
    }

    writeUShort(uShort) {
        this.#buffer.writeUInt16LE(uShort)
    }

    writeInt(int) {
        this.#buffer.writeInt32LE(int)
    }

    writeUInt(uInt) {
        this.#buffer.writeUInt32LE(uInt)
    }

    writeLong(long) {
        this.#buffer.writeBigInt64LE(long)
    }

    writeULong(uLong) {
        this.#buffer.writeBigUInt64LE(uLong)
    }

    writeFloat(float) {
        this.#buffer.writeFloatLE(float)
    }

    writeDouble(double) {
        this.#buffer.writeDoubleLE(double)
    }

    writeBuffer(buffer) {
        this.#buffer.writeBuffer(buffer)
    }

    writeString(string) {
        this.#buffer.writeString(string)
    }

    writeNtString(string) {
        this.#buffer.writeStringNT(string)
    }

    toBuffer() {
        this.#flushBits()
        return this.#buffer.toBuffer()
    }
}

export default SerializerBuffer

File: kotlin/buffela/serialization/src/commonMain/kotlin/serialization/SerializerBuffer.kt
package gr.elaevents.buffela.serialization

import gr.elaevents.buffela.utils.LinkedList
import kotlinx.io.Buffer
import kotlinx.io.readByteArray
import kotlinx.io.writeDoubleLe
import kotlinx.io.writeIntLe
import kotlinx.io.writeLongLe
import kotlinx.io.writeShortLe
import kotlinx.io.writeString
import kotlinx.io.writeUByte
import kotlinx.io.writeUIntLe
import kotlinx.io.writeULongLe
import kotlinx.io.writeUShortLe

private data class BitChunk(val offset: Int, var buffer: Byte = 0) {
    fun apply(bytes: ByteArray) {
        bytes[offset] = buffer
    }
}

class SerializerBuffer {
    private val buffer = Buffer()
    private val bitChunks = LinkedList<BitChunk>()

    private var bitChunk: BitChunk? = null
    private var bitBuffer: Int = 0
    private var bitCount = 0

    val length get() = buffer.size.toInt()

    private fun flushBits() {
        if (this.bitCount == 0) return
        bitChunk!!.buffer = this.bitBuffer.toByte()
    }

    private fun writeLSBits(value: Int, bitLength: Int) {
        if (this.bitCount == 0) {
            this.bitChunk = BitChunk(this.length)
            bitChunks.append(this.bitChunk!!)

            this.buffer.writeByte(0) // Reserve space
        }

        val mask = (1 shl bitLength) - 1
        val masked = mask and value
        val shifted = masked shl this.bitCount

        this.bitCount += bitLength
        this.bitBuffer = this.bitBuffer or shifted
    }

    private fun writeTruncated(value: Int, bitLength: Int) {
        var remainingValue = value
        var remainingLength = bitLength

        while (remainingLength > 0) {
            val available = 8 - this.bitCount
            if (remainingLength < available) {
                this.writeLSBits(remainingValue, remainingLength)
                break;
            }

            this.writeLSBits(remainingValue, available)
            this.flushBits()
            this.clearBitBuffer()

            remainingValue = remainingValue ushr available
            remainingLength -= available
        }
    }

    fun clearBitBuffer() {
        this.bitCount = 0
        this.bitBuffer = 0
    }

    fun writeSigned(value: Int, bitLength: Int) {
        if (bitLength > 31) throw IllegalArgumentException("Bit fields cannot be larger than 31 bits")

        val minValue = -(1 shl bitLength - 1)
        val maxValue = (1 shl bitLength) - 1

        if (value !in minValue..maxValue)
            throw IllegalArgumentException("Value out of range")

        this.writeTruncated(value, bitLength)
    }

    fun writeUnsigned(value: UInt, bitLength: Int) {
        if (bitLength > 31) throw IllegalArgumentException("Bit fields cannot be larger than 31 bits")

        val maxValue = (1u shl bitLength) - 1u
        if (value !in 0u..maxValue)
            throw IllegalArgumentException("Value out of range")

        this.writeTruncated(value.toInt(), bitLength)
    }

    fun writeByte(byte: Byte) {
        this.buffer.writeByte(byte)
    }

    fun writeUByte(uByte: UByte) {
        this.buffer.writeUByte(uByte)
    }

    fun writeShort(short: Short) {
        this.buffer.writeShortLe(short)
    }

    fun writeUShort(uShort: UShort) {
        this.buffer.writeUShortLe(uShort)
    }

    fun writeInt(int: Int) {
        this.buffer.writeIntLe(int)
    }

    fun writeUInt(uInt: UInt) {
        this.buffer.writeUIntLe(uInt)
    }

    fun writeLong(long: Long) {
        this.buffer.writeLongLe(long)
    }

    fun writeULong(uLong: ULong) {
        this.buffer.writeULongLe(uLong)
    }

    fun writeDouble(double: Double) {
        this.buffer.writeDoubleLe(double)
    }

    fun writeBoolean(boolean: Boolean) {
        this.writeTruncated(if (boolean) 1 else 0, 1)
    }

    fun writeBytes(bytes: ByteArray) {
        this.buffer.write(bytes, 0, bytes.size)
    }

    fun writeString(string: String, nt: Boolean = false) {
        this.buffer.writeString(string)
        if (nt) this.buffer.writeByte(0)
    }

    fun toBytes(): ByteArray {
        this.flushBits()

        // Make copy so that we don't consume the bytes from the buffer
        val bytes = this.buffer.copy().readByteArray()

        for (chunk in this.bitChunks)
            chunk.apply(bytes)

        return bytes
    }
}